### PR TITLE
Customization field multishop

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Handle multishop for customization field in product V2, the name can be edited differently on each shop however we don't handle the `Apply to all shops` checkbox
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | ~
| Related PRs       | ~
| How to test?      | This PR handles the multishop feature for customization fields in product page V2, so you should focus testing on these fields which have following rules in multishop:<br>- the creation and deletion of a customization field is common to ALL SHOPS, all the shops share the same number of fields<br>- the type and require values of fields are also common to ALL shops<br>- the only field that you can change depending on the shop is its name<br>- in product page V2 you can edit the name field differently by switching from a shop to another<br>- the `Modify for all shops` checkbox feature IS NOT implemented for this field
| Possible impacts? | ~

### CODE REVIEW

File: src/Adapter/Product/Customization/Repository/CustomizationFieldRepository.php [hunk 1]
```diff
@@ -29,35 +29,57 @@
...
         CustomizationFieldValidator $customizationFieldValidator
     ) {
+        $this->connection = $connection;
+        $this->dbPrefix = $dbPrefix;
         $this->customizationFieldValidator = $customizationFieldValidator;
     }
 
     /**
+     * This getter without specified shop is useful when the product is only fetched for deletion.
+     * In which case the shopId doesn't matter e don't care about multishop content since the entity is about to be deleted.
```
Comment (matks) [inline src/Adapter/Product/Customization/Repository/CustomizationFieldRepository.php]:
```suggestion
     * In which case the shopId doesn't matter we don't care about multishop content since the entity is about to be deleted.
```

**Review by matks — DISMISSED** [review]

File: src/Adapter/Product/Customization/Repository/CustomizationFieldRepository.php [hunk 2]
```diff
@@ -118,4 +163,31 @@ public function softDelete(CustomizationField $customizationField): void
...
     }
+
+    /**
+     * Returns the list of customization associated to a product (only their IDs), by default soft deleted entities
+     * are filtered.
+     *
+     * @param ProductId $productId
+     * @param bool $includeSoftDeleted
+     *
+     * @return array
```
Comment (zuk3975) [inline src/Adapter/Product/Customization/Repository/CustomizationFieldRepository.php]: psalm annotation missing

Comment (zuk3975) [reply src/Adapter/Product/Customization/Repository/CustomizationFieldRepository.php]:
```suggestion
     * @return CustomizationFIeldId[]
```

File: src/Adapter/Product/Customization/Update/ProductCustomizationFieldUpdater.php [hunk 3]
```diff
@@ -75,30 +75,37 @@ public function __construct(
...
 
     /**
-     * @param Product $product
+     * @param ProductId $productId
      */
-    public function refreshProductCustomizability(Product $product): void
+    public function refreshProductCustomizability(ProductId $productId): void
     {
+        // Although the modified fields are defined as multishop, they depend on the association with customization
+        // fields which are not, so all those fields must be synced on all associated shop
```
Comment (zuk3975) [inline src/Adapter/Product/Customization/Update/ProductCustomizationFieldUpdater.php]: almost the same, but I think its clearer (was abit hard to understand from first look)
```suggestion
        // The modified fields are defined as multishop, but they depend on the association with customization fields
        // which are not multishop, so all those fields must be synced between all associated shops
```

**Review by zuk3975 — APPROVED** [review]

File: src/Core/Domain/Product/Customization/Command/SetProductCustomizationFieldsCommand.php [hunk 4]
```diff
@@ -47,14 +51,24 @@ class SetProductCustomizationFieldsCommand
...
     private $customizationFields = [];
 
+    /**
+     * @var ShopConstraint
+     */
+    private $shopConstraint;
+
     /**
      * @param int $productId
      * @param array $customizationFields
```
Comment (zuk3975) [inline src/Core/Domain/Product/Customization/Command/SetProductCustomizationFieldsCommand.php]: could you define array structure here?

File: src/Core/Domain/Product/Customization/CustomizationShopConstraintTrait.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * A POC had been started in case the feature needs to evolve someday https://github.com/PrestaShop/PrestaShop/pull/27944
+ */
+trait CustomizationShopConstraintTrait
+{
+    /**
+     * @param ShopConstraint $shopConstraint
+     *
+     * @throws InvalidShopConstraintException
+     */
+    protected function checkShopConstraint(ShopCOnstraint $shopConstraint): void
```
Comment (zuk3975) [inline src/Core/Domain/Product/Customization/CustomizationShopConstraintTrait.php]:
```suggestion
    protected function checkShopConstraint(ShopConstraint $shopConstraint): void
```

Comment (zuk3975) [reply src/Core/Domain/Product/Customization/CustomizationShopConstraintTrait.php]: :smile:

File: tests/Integration/Behaviour/Features/Context/Domain/Product/UpdateCustomizationFieldsFeatureContext.php [hunk 6]
```diff
@@ -241,20 +328,44 @@ public function assertCustomizationFields(string $productReference, TableNode $t
...
 
-    /**
-     * @Then product :productReference should have :expectedCount customizable :customizationType field(s)
-     *
-     * @param string $productReference
-     * @param int $expectedCount
-     * @param string $customizationType
-     */
-    public function assertCustomizationOptions(string $productReference, int $expectedCount, string $customizationType)
+    private function assertCustomizability(string $productReference, string $customizability, int $shopId)
```
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/UpdateCustomizationFieldsFeatureContext.php]:
```suggestion
    private function assertCustomizability(string $productReference, string $customizability, int $shopId): void
```

File: tests/Integration/Behaviour/Features/Context/Domain/Product/UpdateCustomizationFieldsFeatureContext.php [hunk 7]
```diff
@@ -241,20 +328,44 @@ public function assertCustomizationFields(string $productReference, TableNode $t
...
+                    sprintf('Expected product "%s" to require customization', $productReference)
+                );
+
+                break;
+            default:
+                throw new RuntimeException(sprintf('Invalid customizability "%s" provided in test scenario', $customizability));
+        }
+    }
+
+    private function assertCustomizationOptions(string $productReference, int $expectedCount, string $customizationType, int $shopId)
```
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/UpdateCustomizationFieldsFeatureContext.php]:
```suggestion
    private function assertCustomizationOptions(string $productReference, int $expectedCount, string $customizationType, int $shopId); void
```

**Review by zuk3975 — CHANGES_REQUESTED** [review]
Some minor fixes needed

**Review by zuk3975 — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Comment by AureRita** [comment]
Hi @jolelievre 

Thank you for your PR, I tested it and the main feature works, even the modify for all shops checkbox, as we can see here :

[Untitled_ Nov 28, 2022 2_43 PM.webm](https://user-images.githubusercontent.com/100699445/204293667-cd4b5dd0-8b64-439c-92f4-b07d82307f37.webm)

But after checking some other tab I can see that we have some issues :

 on the UI as you can see.  

image

---------------------------------

With combination ( we didn't check the modify for all shop and the reference change for all. Morevover, when we put the price for all, that didn't change for all but this one is maybe not implemented currently ) : 

[Untitled_ Nov 28, 2022 2_55 PM.webm](https://user-images.githubusercontent.com/100699445/204295210-44065c6e-0ed9-4a19-b48b-035514c70e89.webm)

---------------------------------
With tags (when you add a tags, it should be on the two shop and it seems to be only on the shop where you add it) 

[Untitled_ Nov 28, 2022 3_32 PM.webm](https://user-images.githubusercontent.com/100699445/204303936-215814dd-e783-4039-ad10-55ff59460b9b.webm)

-----------------------------------
With visibility (when you disable the view on the catalog or in the search it seems to disabled it on all shop but when you enable, no) 

[Untitled_ Nov 28, 2022 3_48 PM.webm](https://user-images.githubusercontent.com/100699445/204307393-1e061d5a-0a2e-49da-b9bd-117c2ce5b01f.webm)

Can you say me if one of them it's correct ? 

Waiting for your feedback

**Comment by jolelievre** [comment]
Thanks @AureRita 

all your feedbacks are not in the scope of this PR at all.

I can still answer them though:
- combination multidhop edition has not been implemented yet so it's bound to have some issues
- for the tags we didn't focus on it yet, but I didn't think it was needed As you said the tags are common to all shops so if you add tags on a shop it will impact all the others I don't understand you bug report though because in your video it seems like the tag is indeed in the second shop so no bug in the BO apparently Or do you mean it doesn't work on the FO when you search the tag on the second shop? If so then please open a dedicated issue for this bug
- about the visibility, the for in the BO behaves as expected at least but it seems like the FO doesn't react appropriately I'm not sure if it's a bug from the BO (which doesn't index the products properly maybe) or from the FO (and maybe it's already an existing bug, did you test with page V1) But anyway it's not related to this PR either so it deserves a dedicated issue as well

TLDR; if you didn't find bugs related to customization field then I guess this PR is validated.

**Review by florine2623 — APPROVED** [review]
Hello @jolelievre ,

Tested on `develop`, check the customization field in product edition in PPV2.

Tested on different configurations of multishop:
- multishop disabled 
- multishop enabled 
- multishop enabled, multiple shops in 1 group
- multishop enabled, multiple shops in multiple group (some shops configured, some not)

Checked:
- with 1 customization, multiple customizations
- type file and type text
- required or not
- on different types of product (standard, combination, pack, virtual)

**Issues found:**

When adding a file in FO that is not .gif .jpg .jpeg .jpe .png .webp, Fatal error: :x:

https://user-images.githubusercontent.com/16019289/205095346-ca0d4a4b-a763-49c2-9b92-eefad63c2ea6.mov

It is a regression on 8.0.0, it should be like this:
image (3)


Notice is displayed when non-required customization fields are saved: :x:
Issue created by @paulnoelcholot : https://github.com/PrestaShop/PrestaShop/issues/30464

https://user-images.githubusercontent.com/16019289/205096328-1ef75284-cb3f-4e68-8cb2-b8d987923fc1.mov


**Second part:**

_I have also found the following, but I am unable to reproduce them on another shop with PR with the same steps. I'll have to have a look later._

- Exception displayed in alert:
Screenshot 2022-12-01 at 16 29 33

- Click on Select shops on the top-right corner, see that the shop info is not coherent
Screenshot 2022-12-01 at 16 30 44

- Display of the popup when creating a product is not shown correctly:
Screenshot 2022-12-01 at 16 28 09

It should be like this:
Screenshot 2022-12-01 at 16 51 40

I have also found this exception on one of my product, I couldn't find the root cause of it :/

Screenshot 2022-12-01 at 17 12 34

https://user-images.githubusercontent.com/16019289/205105312-f97a599f-5ad3-41e3-95c4-281b3a04b49d.mov

Error message in alert is not clear. Even after adding a title and a refresh of the page, the message is not updated

https://user-images.githubusercontent.com/16019289/205106303-4ffe2f27-f190-47cc-880c-5dd6683a17f7.mov


I think all the bugs mentioned are non-blocking or not in this scope. So I validate the PR.
Please make other PRs for the 2 bugs that are easy to reproduce. 🙏 
For the second part, I'm sorry, I will check further and let you know if I can have some clear reproducible steps.

It is QA ✅

**Comment by nicosomb** [comment]
Thank you @jolelievre !
